Add v8 migration implementation and comprehensive documentation

hkollmann opened the pull request:
## Summary

This PR implements a comprehensive automatic migration system for upgrading Qooxdoo applications from v7 to v8, along with complete documentation of all breaking changes.

## Changes

### Migration Implementation (`source/class/qx/tool/migration/M8_0_0.js`)

Created a new migration class with 10 migration methods covering all v8.0.0_beta breaking changes:

1. **migrateManifest()** - Updates Manifest.json and @qooxdoo/framework dependency
2. **migrateTableModelUsage()** - Warns about qx.ui.table.Table editing constraint
3. **migrateDeprecatedAPIs()** - Scans for deprecated API usage
4. **migrateCompileJson()** - Updates compile.json schema
5. **migrateCompileJs()** - Detects and provides migration guide for yargs → CLI classes
6. **migrateInstanceName()** - **Automatically** replaces `.name` → `.classname`
7. **migratePropertyMemberConflicts()** - Warns about property/member namespace unification
8. **migrateNodeVersion()** - Warns about Node.js >= 20.0.0 requirement and ESLint 9
9. **migrateLocaleAPI()** - Warns about qx.locale CLDR → Intl API changes
10. **migratePackages()** - Upgrades packages to v8-compatible versions

### Documentation

- **`docs/release-notes-8.0.md`** - Enhanced with:
  - All breaking changes from CHANGELOG.md
  - Code examples for CLI migration (yargs → CLI classes)
  - ESLint 9 Flat Config details
  - Complete migration guide section with:
    - Automatic migrations performed
    - 7-point manual review checklist
    - Step-by-step execution instructions
    - Dry-run and verbose mode documentation

## Breaking Changes Covered

1. ✅ CLI System Migration (yargs → CLI classes)
2. ✅ Properties and Members in Same Namespace
3. ✅ Property Refining Behavior
4. ✅ instance.name → instance.classname (automatic)
5. ✅ Property Check as String No Longer Supported
6. ✅ ESLint 8 → ESLint 9 (Node.js >= 20.0.0)
7. ✅ qx.locale CLDR → Intl API
8. ✅ qx.ui.table.Table Editing Constraint
9. ✅ Appearance and Font Color Behavior
10. ✅ qx.util.PropertyUtil API Changes

## Migration Features

### Automatic
- Manifest.json updates
- instance.name → classname replacement
- Package upgrades

### Warnings with Guidance
- CLI system changes with code examples
- Property/member conflicts
- Node.js version requirements
- ESLint 9 migration info
- Locale API changes
- Table model usage
- Deprecated APIs

## Usage

```bash
# Preview changes
qx migrate --dry-run --verbose

# Execute migration
qx migrate --verbose

Commit:
fix api doc

claude committed:
Add comprehensive v8 migration implementation and documentation

This commit adds complete migration support for Qooxdoo v8:

New Migration Class (M8_0_0.js):
- migrateManifest() - Updates Manifest.json and dependencies
- migrateTableModelUsage() - Warns about table editing constraint
- migrateDeprecatedAPIs() - Informs about deprecated APIs
- migrateCompileJson() - Updates compile.json schema
- migrateCompileJs() - Detects yargs usage, provides CLI migration guide
- migrateInstanceName() - Warns about instance.name → classname change
- migratePropertyMemberConflicts() - Warns about namespace unification
- migrateNodeVersion() - Node.js >= 20.0.0 and ESLint 9 warnings
- migrateLocaleAPI() - CLDR → Intl API change information
- migrateFormAddNames() - **Intelligent scanner** for Form.add() uppercase names
  - Recursively scans all source files
  - Detects .add() calls with uppercase name parameters
  - Reports specific files and suggests corrections
- migratePackages() - Upgrades packages to v8-compatible versions

Release Notes Updates (release-notes-8.0.md):
- Added "Notable changes and new features in v8.0" section with:
  - New CLI Architecture
  - ESLint 9 Support with Flat Config
  - Significantly Reduced Package Size (qx.locale)
  - Automatic Migration Tool overview
  - Improved Developer Experience
- Enhanced "Breaking changes in v8.0" with:
  - CLI System Migration (yargs → CLI classes) with code examples
  - qx.ui.table.Table editing constraint
  - ESLint 8 → ESLint 9 migration details
  - qx.locale CLDR → Intl API changes
  - qx.util.PropertyUtil API changes
  - qx.ui.core.MExecutable binding changes
  - **Form.add() lowercase requirement** with examples
- Added "Migration from v7 to v8" section with:
  - Automatic migrations list
  - 9-point manual review checklist
  - Step-by-step migration instructions
  - Dry-run and verbose mode documentation

Key Features:
✅ 11 migration methods covering all documented breaking changes
✅ Intelligent Form.add() scanner with file-specific reports
✅ Comprehensive documentation in release notes
✅ Clear migration guide for users
✅ Supports dry-run and verbose modes

claude committed:
Add class name migration detection to migrateCompileJs

Extended migrateCompileJs to detect and warn about renamed class names in v8:

Class name changes:
- qx.tool.cli.commands.* → qx.tool.compiler.cli.commands.*
- qx.tool.cli.ConfigDb → qx.tool.compiler.cli.ConfigDb
- qx.tool.cli.commands.Package → qx.tool.compiler.cli.commands.Package

Features:
- Detects old class names in compile.js
- Provides clear OLD → NEW mapping for each found class name
- Combines with existing yargs detection
- Shows all warnings together for comprehensive migration guide

Example output when old class names found:
  "OLD: qx.tool.cli.commands.
   NEW: qx.tool.compiler.cli.commands.

   OLD: qx.tool.cli.ConfigDb
   NEW: qx.tool.compiler.cli.ConfigDb"

This helps users migrating their compile.js files from v7 to v8.

claude committed:
Improve migrateCompileJs: automatic class name updates, separate yargs warning

Commit:
added more notes

hkollmann committed:
fix lint

hkollmann committed:
Merge branch 'qooxdoo:branch_8_0_beta' into branch_8_0_beta

hkollmann committed:
add comprehensive test suite for M8_0_0 migration

Created test-qx-migrate.js with 37 tests (100% passing) validating all M8_0_0
migration functionality including Manifest updates, compile.js class name
replacements, form.add() uppercase detection, and breaking change announcements.

Bug fixes discovered during test development:
- BaseMigration.js: Fix checkFilesContain() to handle both RegExp and strings
- M8_0_0.js: Skip package upgrade when contrib.json has no libraries

Updated test-migrations.js to support partial migrations where some migrations
are auto-applied while others require manual review (v8.0.0 pattern: 4 applied,
5 pending).

Added v8.0.0 test fixtures with minimal but complete project structure.

derrell commented on docs/release-notes-8.0.md:
```diff
@@ -147,12 +223,234 @@
   may be other obscure backward-compatibility changes that pop up.
   Those listed above are the only ones that reared their heads while
   confirming that the entire qooxdoo test suite successfully runs to
   completion.
```
I didn't see any mention of the fact that property names and members are now in the same namespace, so any classes which defined both a property with name `x` and a member variable with name `x` will need to change the name of one of them. (Does `qx migrate` identify and/or modify code to correct for this?)

derrell commented on source/class/qx/tool/migration/M8_0_0.js:
```diff
@@ -0,0 +1,378 @@
+        "Also note: Refining a property in a subclass now adds it to the\n" +
+        "subclass prototype instead of modifying it in place."
+      );
+      this.markAsPending("Manual review of property/member conflicts required");
```
This seems inadequate. Unlike most of the other migrations, where a search for something specific such as `\.name` is pretty easy to accomplish, manually searching each and every property name in every class for a member variable of the same name (or vice versa) is a LOT of work (nearly impossible), whereas programmatically doing it should be relatively easy. This method should be implemented to automatically search for and alert for any such conflicts.

hkollmann committed:
looks nicer

hkollmann committed:
https://github.com/qooxdoo/qooxdoo/pull/10830#discussion_r2628524019

hkollmann committed:
https://github.com/qooxdoo/qooxdoo/pull/10830#discussion_r2628524019

hkollmann committed:
re add library loading
fix loading of old libs
fix api for compile.js

hkollmann committed:
@derrel:
This seems inadequate. Unlike most of the other migrations, where a search for something specific such as \.name is pretty easy to accomplish, manually searching each and every property name in every class for a member variable of the same name (or vice versa) is a LOT of work (nearly impossible), whereas programmatically doing it should be relatively easy. This method should be implemented to automatically search for and alert for any such conflicts.

hkollmann committed:
Add migration check for constructor property setter order

In v8, property setters must be called AFTER super() or this.base().
Values set before the parent constructor will be reset when the parent
constructor executes.

This migration check:
- Scans constructor methods in qx.Class.define and qx.Mixin.define
- Detects this.setXxx() calls before super() or this.base()
- Reports file paths, line numbers, and specific setter methods
- Provides clear guidance on how to fix the issue

Example broken pattern (v7):
  construct(width, height) {
    this.setWidth(width);   // Will be lost!
    this.setHeight(height);
    this.base(arguments);
  }

Correct pattern (v8):
  construct(width, height) {
    this.base(arguments);   // Call parent first
    this.setWidth(width);   // Now set properties
    this.setHeight(height);
  }

Implementation:
- Added migrateConstructorPropertySetters() method to M8_0_0
- Added __findSettersBeforeSuper() AST helper
- Integrated with unified __scanSourceFilesForIssues() scanner
- Added test fixtures: ConstructorSetterTest, NoConstructorSetterTest
- Added integration test with 11 assertions (all passing)

hkollmann committed:
Add qx_packages, node_modules, and qx-lock.json to test migration .gitignore

These files are generated during test execution and should not be tracked.

hkollmann committed:
fix simple test for v8

oetiker reviewed (approved):
wow!
